### PR TITLE
Frontend author/hub privilege check for events/posts

### PR DESCRIPTION
**Pull request by nnguyen144**
### Category
- [ ] **Feature**
- [X] **Bug Fix**
- [ ] **Testing**
- [ ] **Documentation**
- [ ] **Other**

### Domain
- [ ] **Frontend**
- [ ] **Backend**
- [X] **Both**

### Description
This PR adds helper functions that check if the current user is an author of or is a moderator/owner of a hub for an event or post.

### Related Issue(s)
[https://github.com/UNLV-CS472-672/2024-F-GROUP3-RebelHub/issues/47](https://github.com/UNLV-CS472-672/2024-F-GROUP3-RebelHub/issues/47)

### New Additions
1 new file was added: fetchPrivileges.js
This file includes the helper functions that check author and hub privileges.

### Modifications
Files modified: 
backend/calendar_app/serializers.py: added author to EventCreateSerializer field so that I can retrieve the author from newly created posts without having to refresh the browser.
backend/calendar_app/views.py: removed unnecessary import.
backend/calendar_app/serializers.py: added id to UserSerializer as a read-only field so that I can retrieve user id from api in order to compare with "author" of an event/post.
frontend/src/components/Calendar/Calendar.js: updated to use the new helper functions
frontend/src/components/Calendar/EventModal.js: updated to use the new helper functions
frontend/src/components/posts/post-summary.tsx: updated to use the new helper functions to hide delete button for posts where the user does not have the privilege to delete.

### Architecture of Your Changes
fetchPrivileges.js has four functions:
fetchHubs(): This function fetches the hubs from the api that the current user is moderating/owns.
fetchHubsIDs(): This function calls fetchHubs() and retrieves the IDs from the list of Hub objects.
checkHubPrivileges(hub_id): This function calls fetchHubsIDs() and returns true if hub_id (representing the id of the hub for an event/post) is included in the list of hub IDs.
checkAuthorPrivileges(author): This function fetches the current user from the api and checks if the author id of an event/post matches with the current user.

### Usage and Integration
These helper functions can be included in the frontend for events, posts, and potentially anything else in the frontend that requires checking for author and hub privileges.

### Testing
-**Manual Testing**: I've tested the code manually and witnessed that the results of the code are outputted correctly, meaning that if the current user is neither the author of an event/post nor a moderator/owner of the hub of the event/post, then I should not see the following:

1. Delete button for posts.
2. Update and delete button for events.

The fetchPrivileges.js file's fetchHubs() function is also used when I display the hubs that the user owns/moderates when creating an event in CreateForm.js.

### CODE REVIEW

**Review by leed85 — APPROVED** [review]

**Comment by thecatfour** [comment]
When I was testing this PR, I found that I could not delete posts as a moderator for a hub. This appears to be a backend problem for posts as the button appears correctly. The following is the steps you can take to recreate this problem:

1. Create an owner
2. Create a moderator
3. Have owner create a hub
4. Make moderator a moderator in the hub
5. Create a user
6. Have user create a post
7. Try to delete the post as moderator
8. Get error 403

I believe the issue is in backend/Posts/views.py in the class PostDelete. I fixed this issue by adding the following if statement to the perform_destroy function:
```
def perform_destroy(...):
   ...
   if user != instance.author or user != instance.hub.owner:
      if user not in instance.hub.mods.all():
         raise PermissionDenied()
   instance.delete()
```
Basically, I just add a check to see if the user is a moderator for the hub the post was created in, and if they are, then allow the delete to go through.

**Comment by thecatfour** [comment]
Additionally, I ran into a runtime error when selecting a hub for the create a event modal on the Calendar page. This seems to just be a renaming parameters issue though. The "hubsModding" and "hubs" parameters do not match, which makes the parameter undefined in the modal rather than an empty list.

image

image

Basically, you changed the name of the parameter in Calendar.js without changing the name of the parameter in the CreateForm.js component. I recommend keeping the original hubs name or refactoring CreateForm with the new name.

**Comment by nnguyen144** [comment]
Thanks for the recommendations, I've gone ahead and implemented them.

**Comment by thecatfour** [comment]
When talking about the bug for PostDelete, I combined the instance.author and instance.hub.owner into one statement to highlight the if statement that I added. I did not mean to imply that you should change those lines. I will be more clear in the future.

**Review by thecatfour — APPROVED** [review]
I believe that the PR accomplishes its intended goal now. It would be a good idea in the future to create some tests in Posts to check if moderators can delete posts as that is part of the role of moderators.

Edit: I created issue #61 to highlight the new moderator tests that we should implement.
